File: routers/module_types.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy import func
from sqlmodel import Session, select
from typing import Dict, List, Optional

from database import get_session
from models import Module, ModuleTypeDefinition
from schemas import ModuleTypeDefinitionCreate, ModuleTypeDefinitionRead, ModuleTypeDefinitionUpdate
from tenancy import CurrentOrg

router = APIRouter()

BUILTIN_TYPES = [
    {"key": "breaker",          "name_no": "Automatsikring",     "color": "#2563eb", "abbreviation": "LS", "can_have_circuit": True,  "can_have_ampere": True},
    {"key": "rcd",              "name_no": "Jordfeilbryter",     "color": "#ca8a04", "abbreviation": "JF", "can_have_circuit": False, "can_have_ampere": True},
    {"key": "rcd_breaker",      "name_no": "Kombibryter",        "color": "#16a34a", "abbreviation": "KO", "can_have_circuit": True,  "can_have_ampere": True},
    {"key": "shelly",           "name_no": "Shelly",             "color": "#ea580c", "abbreviation": "SH", "can_have_circuit": False, "can_have_ampere": False},
    {"key": "dynalite",         "name_no": "Dynalite",           "color": "#9333ea", "abbreviation": "DY", "can_have_circuit": False, "can_have_ampere": False},
    {"key": "surge_protection", "name_no": "Overspenningsvern",  "color": "#dc2626", "abbreviation": "OV", "can_have_circuit": False, "can_have_ampere": False},
    {"key": "main_switch",      "name_no": "Hovedbryter (OV50)", "color": "#374151", "abbreviation": "OV", "can_have_circuit": False, "can_have_ampere": True},
    {"key": "other",            "name_no": "Annet",              "color": "#6b7280", "abbreviation": "—",  "can_have_circuit": False, "can_have_ampere": False},
]


def seed_builtin_types(session: Session) -> None:
    """Insert any built-in type this database does not have yet.

    Built-ins carry organization_id = NULL and are shared by every tenant.
    Upserting per key rather than bailing out when the table is non-empty:
    the old version returned early if any row existed, so a type added to
    BUILTIN_TYPES later never reached an existing install.
    """
    existing = {
        row.key
        for row in session.exec(
            select(ModuleTypeDefinition).where(ModuleTypeDefinition.is_builtin == True)  # noqa: E712
        ).all()
    }
    added = False
    for data in BUILTIN_TYPES:
        if data["key"] in existing:
            continue
        session.add(ModuleTypeDefinition(is_builtin=True, organization_id=None, **data))
        added = True
    if added:
        session.commit()


def _usage_counts(session: Session, org_id: int) -> Dict[str, int]:
    """How many modules use each type, in one query rather than one per type."""
    rows = session.exec(
        select(Module.type, func.count(Module.id))
        .where(Module.organization_id == org_id)
        .group_by(Module.type)
    ).all()
    return {key: count for key, count in rows}


def _visible(session: Session, org_id: int) -> List[ModuleTypeDefinition]:
    """Built-in types plus this organization's own.

    Where an organization has customised a built-in, its version shadows the
    shared one — same key, one entry. See update_module_type.
    """
    rows = session.exec(
        select(ModuleTypeDefinition).where(
            (ModuleTypeDefinition.organization_id == org_id)
            | (ModuleTypeDefinition.organization_id == None)  # noqa: E711
        )
    ).all()
    by_key: Dict[str, ModuleTypeDefinition] = {}
    for row in rows:
        if row.key not in by_key or row.organization_id is not None:
            by_key[row.key] = row
    return list(by_key.values())


def _read(mtd: ModuleTypeDefinition, counts: Dict[str, int]) -> ModuleTypeDefinitionRead:
    d = ModuleTypeDefinitionRead.model_validate(mtd)
    d.usage_count = counts.get(mtd.key, 0)
    return d


def _get_visible(type_id: int, session: Session, org_id: int) -> ModuleTypeDefinition:
    mtd = session.get(ModuleTypeDefinition, type_id)
    if not mtd or mtd.organization_id not in (org_id, None):
        raise HTTPException(status_code=404, detail="Module type not found")
    return mtd


@router.get("", response_model=List[ModuleTypeDefinitionRead])
def list_module_types(org: CurrentOrg, session: Session = Depends(get_session)):
    counts = _usage_counts(session, org)
    result = [_read(row, counts) for row in _visible(session, org)]
    return sorted(result, key=lambda r: (0 if r.is_builtin else 1, r.name_no))


@router.get("/{type_id}", response_model=ModuleTypeDefinitionRead)
def get_module_type(type_id: int, org: CurrentOrg, session: Session = Depends(get_session)):
    mtd = _get_visible(type_id, session, org)
    return _read(mtd, _usage_counts(session, org))


@router.post("", response_model=ModuleTypeDefinitionRead)
def create_module_type(
    data: ModuleTypeDefinitionCreate, org: CurrentOrg, session: Session = Depends(get_session)
):
    # A key must be unique among what this organization can see, which
    # includes the shared built-ins.
    if any(row.key == data.key for row in _visible(session, org)):
        raise HTTPException(status_code=400, detail="Module type key already exists")
    mtd = ModuleTypeDefinition(is_builtin=False, organization_id=org, **data.model_dump())
    session.add(mtd)
    session.commit()
    session.refresh(mtd)
    return _read(mtd, _usage_counts(session, org))


@router.put("/{type_id}", response_model=ModuleTypeDefinitionRead)
def update_module_type(
    type_id: int,
    data: ModuleTypeDefinitionUpdate,
    org: CurrentOrg,
    session: Session = Depends(get_session),
):
    mtd = _get_visible(type_id, session, org)
    changes = data.model_dump(exclude_unset=True)

    if mtd.organization_id is None:
        # Built-ins are shared by every tenant, so editing one in place would
        # change it for everyone. Copy on write instead: this organization
        # gets its own version, which shadows the shared one by key. Deleting
        # that copy later reverts to the default.
        mtd = ModuleTypeDefinition(
            organization_id=org,
            key=mtd.key,
            is_builtin=True,
            name_no=changes.get("name_no", mtd.name_no),
            color=changes.get("color", mtd.color),
            abbreviation=changes.get("abbreviation", mtd.abbreviation),
            can_have_circuit=changes.get("can_have_circuit", mtd.can_have_circuit),
            can_have_ampere=changes.get("can_have_ampere", mtd.can_have_ampere),
        )
    else:
        for field, value in changes.items():
            setattr(mtd, field, value)

    session.add(mtd)
    session.commit()
    session.refresh(mtd)
    return _read(mtd, _usage_counts(session, org))


@router.delete("/{type_id}", response_model=ModuleTypeDefinitionRead)
def delete_module_type(type_id: int, org: CurrentOrg, session: Session = Depends(get_session)):
    mtd = _get_visible(type_id, session, org)
    if mtd.organization_id is None:
        # Shared across every tenant — not this organization's to remove.
        raise HTTPException(status_code=409, detail="Built-in types cannot be deleted")
    counts = _usage_counts(session, org)
    count = counts.get(mtd.key, 0)
    if count > 0:
        raise HTTPException(
            status_code=409,
            detail=f"Cannot delete: {count} module(s) use this type",
        )
    snapshot = ModuleTypeDefinitionRead.model_validate(mtd)
    snapshot.usage_count = 0
    session.delete(mtd)
    session.commit()
    return snapshot


@router.get("/{key}/usage")
def get_module_type_usage(key: str, org: CurrentOrg, session: Session = Depends(get_session)):
    return {"key": key, "count": _usage_counts(session, org).get(key, 0)}

File: workers/src/routes/moduleTypes.ts
/**
 * Modultyper — what a module in the panel drawing can be.
 *
 * The interesting behaviour is copy-on-write. Eight built-in types are
 * shared by every tenant and carry no organization. Editing one in place
 * would change the colour of everybody's automatsikring, so an edit makes
 * this organization its own version instead: same key, shadowing the
 * shared one, and deleting it later reverts to the default.
 *
 * That shadowing is why the listing is assembled rather than selected. A
 * customised built-in must appear once, not twice.
 *
 * Nothing here touches a table directly. The definitions come from
 * src/moduleTypes.ts, which is the one module allowed to work unscoped
 * because this is the one table that crosses tenants by design; the usage
 * counts come through Tenant, because modules do not.
 */

import { Hono } from "hono";

import * as s from "../schema";
import type { Env } from "../context";
import type { Tenant } from "../db";
import { fail, intParam } from "../http";
import * as types from "../moduleTypes";
import { MODULE_TYPE_CREATE, MODULE_TYPE_UPDATE, moduleTypeRead } from "../schemas";
import { readBody } from "../validate";

export const moduleTypeRoutes = new Hono<Env>();

/** How many of this organization's modules use each type. */
const usage = (tenant: Tenant) => tenant.countBy(s.module, s.module.type);

moduleTypeRoutes.get("/", async (c) => {
  const tenant = c.get("tenant");
  const [rows, counts] = await Promise.all([
    types.visible(c.env, tenant.organizationId),
    usage(tenant),
  ]);

  // Built-ins first, then the organization's own, each group by name. A
  // customised built-in keeps is_builtin and sorts with the standard ones,
  // which is where the user expects to find it.
  const listed = rows
    .map((row) => moduleTypeRead(row, counts.get(row.key) ?? 0))
    .sort((a, b) =>
      a.is_builtin === b.is_builtin
        ? a.name_no.localeCompare(b.name_no)
        : a.is_builtin
          ? -1
          : 1,
    );

  return c.json(listed);
});

moduleTypeRoutes.post("/", async (c) => {
  const tenant = c.get("tenant");
  const body = await readBody(c.req.raw, MODULE_TYPE_CREATE);
  const key = body.key as string;

  // Unique among what this organization can see, which includes the shared
  // built-ins — two types with one key would make the panel drawing depend
  // on which row a lookup happened to find.
  const existing = await types.visible(c.env, tenant.organizationId);
  if (existing.some((row) => row.key === key)) {
    throw fail(400, "Module type key already exists");
  }

  const row = await types.create(c.env, tenant.organizationId, {
    key,
    nameNo: body.name_no as string,
    color: body.color as string,
    abbreviation: body.abbreviation as string,
    canHaveCircuit: body.can_have_circuit as boolean,
    canHaveAmpere: body.can_have_ampere as boolean,
    isBuiltin: false,
  });

  return c.json(moduleTypeRead(row, (await usage(tenant)).get(key) ?? 0));
});

moduleTypeRoutes.get("/:type_id", async (c) => {
  const tenant = c.get("tenant");
  const row = await visibleOr404(c.env, tenant, c.req.param("type_id"));
  return c.json(moduleTypeRead(row, (await usage(tenant)).get(row.key) ?? 0));
});

moduleTypeRoutes.put("/:type_id", async (c) => {
  const tenant = c.get("tenant");
  const row = await visibleOr404(c.env, tenant, c.req.param("type_id"));
  const body = await readBody(c.req.raw, MODULE_TYPE_UPDATE, { partial: true });

  const changes: Record<string, unknown> = {};
  if ("name_no" in body) changes.nameNo = body.name_no;
  if ("color" in body) changes.color = body.color;
  if ("abbreviation" in body) changes.abbreviation = body.abbreviation;
  if ("can_have_circuit" in body) changes.canHaveCircuit = body.can_have_circuit;
  if ("can_have_ampere" in body) changes.canHaveAmpere = body.can_have_ampere;

  const updated =
    row.organizationId === null
      ? await types.replaceBuiltin(c.env, tenant.organizationId, row, changes)
      : await types.patch(c.env, row.id, changes);

  return c.json(moduleTypeRead(updated, (await usage(tenant)).get(updated.key) ?? 0));
});

moduleTypeRoutes.delete("/:type_id", async (c) => {
  const tenant = c.get("tenant");
  const row = await visibleOr404(c.env, tenant, c.req.param("type_id"));

  if (row.organizationId === null) {
    throw fail(409, "Built-in types cannot be deleted");
  }

  const count = (await usage(tenant)).get(row.key) ?? 0;
  if (count > 0) {
    // The modules would keep their type string and render as blanks with
    // no way to tell why.
    throw fail(409, `Cannot delete: ${count} module(s) use this type`);
  }

  await types.remove(c.env, row.id);
  return c.json(moduleTypeRead(row, 0));
});

/**
 * Usage by key rather than by id.
 *
 * The panel view asks "may I retire this" before offering the option, and
 * it knows the key, not which row backs it for this organization.
 */
moduleTypeRoutes.get("/:key/usage", async (c) => {
  const key = c.req.param("key");
  const counts = await usage(c.get("tenant"));
  return c.json({ key, count: counts.get(key) ?? 0 });
});

async function visibleOr404(
  env: Env["Bindings"],
  tenant: Tenant,
  raw: string,
): Promise<types.ModuleType> {
  const id = intParam(raw, "type_id");
  const row = await types.findVisible(env, tenant.organizationId, id);
  if (!row) throw fail(404, "Module type not found");
  return row;
}
